nrf_wifi/fw_if/umac_if/src/tx.c

```diff
@@ -1517,6 +1846,69 @@ static int get_ac(unsigned int tid,
 	return map_ac_from_tid(tid);
 }
 
+#ifdef CONFIG_NRF700X_RAWDATA_TX
+enum nrf_wifi_status nrf_wifi_fmac_start_rawpkt_xmit(void *dev_ctx,
+						     unsigned char if_idx,
+						     void *nwb)
+{
+	enum nrf_wifi_status status = NRF_WIFI_STATUS_FAIL;
+	enum nrf_wifi_fmac_tx_status tx_status = NRF_WIFI_FMAC_TX_STATUS_FAIL;
+	struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx = NULL;
+	struct nrf_wifi_fmac_dev_ctx_def *def_dev_ctx = NULL;
+	void *nwb_data = NULL;
+	int ac;
+	int peer_id;
+
+	if (!nwb || !dev_ctx) {
+		goto out;
+	}
+	fmac_dev_ctx = (struct nrf_wifi_fmac_dev_ctx *)dev_ctx;
+	def_dev_ctx = wifi_dev_priv(fmac_dev_ctx);
+
+	/**
+	 * only allow raw packet to be transmitted if interface type allows it
+	 * do not queue the packet if interface type does not allow raw tx
+	 */
+
+	if (!nrf_wifi_util_is_rawpktmode_enabled(def_dev_ctx->vif_ctx[if_idx])) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+				      "%s: raw_packet mode is not enabled",
+				      __func__);
+		goto out;
+	}
+
+	nwb_data = nrf_wifi_osal_nbuf_data_get(fmac_dev_ctx->fpriv->opriv,
+					       nwb);
+	nrf_wifi_osal_mem_cpy(fmac_dev_ctx->fpriv->opriv,
+			      &def_dev_ctx->raw_tx_config,
+			      nwb_data,
+			      sizeof(struct raw_tx_data));
+
+	def_dev_ctx->raw_tx_config.raw_tx_flag = 1;
+	peer_id = MAX_PEERS;
+	ac = def_dev_ctx->raw_tx_config.queue;
+
+	tx_status = nrf_wifi_fmac_tx(fmac_dev_ctx,
+				     if_idx,
+				     nwb,
+				     ac,
+				     peer_id);
+
+	if (tx_status == NRF_WIFI_FMAC_TX_STATUS_FAIL) {
+		nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+				      "%s: Failed to send packet\n",
+				      __func__);
+	}
+
+	/**
+	 * Always silently drop the RAW packet and not send Failure.
+	 * The network stack might think interface is down
+	 */
+out:
+	status = NRF_WIFI_STATUS_SUCCESS;
```

Comment:
may be we should add some stats here for debugging?

Reply:
statistics added